### PR TITLE
Sullenor/testcases

### PR DESCRIPTION
**Pull request by sullenor**
I decided to start working on test-coverage improvement.
ref: https://github.com/webpack-contrib/i18n-webpack-plugin/pull/42

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -69,8 +70,9 @@ class I18nPlugin {
           if (typeof result === 'undefined') {
             let error = this.state.module[__dirname]; // eslint-disable-line no-underscore-dangle
             if (!error) {
-              error = this.state.module[__dirname] = // eslint-disable-line no-underscore-dangle
-              new MissingLocalizationError(this.state.module, param, defaultValue);
+              error = new MissingLocalizationError(this.state.module, param, defaultValue);
+              this.state.module[__dirname] = error; // eslint-disable-line no-underscore-dangle
```
Comment (d3viant0ne) [inline src/index.js]: @sullenor - If you run `npm run webpack-defaults` it should pick up the latest version of `eslint-config-webpack` so you don't need to disable `no-underscore-dangle`

File: test/test-cases.setup.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+      resolve(data);
+    });
+  });
+}
+
+export function requireUncache(modulePath) {
+  delete require.cache[require.resolve(modulePath)];
+  return require(modulePath); // eslint-disable-line import/no-dynamic-require, global-require
+}
```
Comment (d3viant0ne) [inline test/test-cases.setup.js]: Something else to keep in mind, as an organization we are trying to retain all of Jest's functionality while still testing through Webpack.

While I understand this is a WIP, it would be worth taking a look at https://github.com/webpack-contrib/css-loader/tree/new-loader/test - this is the general direction we are taking with all of our test implementations.

It allows execution of tests through webpack to catch any edge cases missed by unit testing while maintaining Jest's snapshots, coverage, incremental compile / tests & async features.

If you want or need a hand just @ me

File: package.json [hunk 3]
```diff
@@ -44,6 +44,7 @@
     "nsp": "^2.6.3",
     "pre-commit": "^1.2.2",
     "standard-version": "^4.0.0",
+    "webpack": "^2.4.1",
```
Comment (d3viant0ne) [inline package.json]: `peerDependency` will be sufficient here. It's acceptable to require a developer to install webpack as a part of the devEnv setup. 

For CI, the Travis build runs off a matrix which installs Webpack to test against including multiple versions for canary / major release beta testing as necessary.

**Comment by d3viant0ne** [comment]
I should have some free time this weekend, let me know if you want me to lend a hand.

File: src/index.js (hunk 1 above)
Comment (sullenor) [inline src/index.js]: hm, still have a problem with it after `npm run webpack-defaults`. Should I add `env: node` into the `.eslintrc`?

**Comment by d3viant0ne** [comment]
That did the trick for the incremental compilation / test execution along with the vscode Jest plugin freaking out :)

**Comment by sullenor** [comment]
@d3viant0ne Thank you for the tips. 
Unfortunately, I don't have much time this weekend to finish the pr, so I decided to focus on the common use cases and would be glad if you can help me with turning the testcases automation into the desired way as in the `css-loader` :)

File: test/test-cases.setup.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  const resolvedEntry = join(__dirname, 'test-cases', entry);
+  const resolvedOutput = join(dirname(resolvedEntry), `${basename(resolvedEntry, '.code.js')}.tmp.js`);
+
+  const compiler = webpack({
+    entry: resolvedEntry,
+    output: {
+      filename: basename(resolvedOutput),
+      path: dirname(resolvedOutput),
+
+      libraryTarget: 'commonjs2',
```
Comment (michael-ciniawsky) [inline test/test-cases.setup.js]: Maybe `config.target: 'node'` instead of `config.output.libraryTarget: 'commonjs2'` here 😛

File: .eslintrc [hunk 5]
```diff
@@ -9,6 +9,7 @@
     "no-param-reassign": 1,
     "prefer-destructuring": 1,
     "no-nested-ternary": 1,
+    "no-underscore-dangle": [ 2, { "allow": ["__dirname"], "allowAfterThis": true } ],
```
Comment (michael-ciniawsky) [inline .eslintrc]: 👍 But better handled with `webpack-defaults`, since the `eslint` config is synced there, please additionally open a PR in `eslint-config-webpack` so the next `webpack-defaults` update is not overriding this

File: test/test-cases.setup.js (hunk 4 above)
Comment (sullenor) [inline test/test-cases.setup.js]: `output.libraryTarget` adds the global `module.exports` for the entire module, so it will be possible to require it. I didn't manage to achieve the similar behavior with the `target` option :(

Comment (michael-ciniawsky) [inline test/test-cases.setup.js]: kk then leave it 👍 😛

File: test/test-cases.setup.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
+import { basename, dirname, join } from 'path';
+import webpack from 'webpack';
+import I18nPlugin from '../src';
+
+export default function processFile(entry, ...pluginOpts) {
+  const resolvedEntry = join(__dirname, 'test-cases', entry);
+  const resolvedOutput = join(dirname(resolvedEntry), `${basename(resolvedEntry, '.code.js')}.tmp.js`);
```
Comment (sullenor) [inline test/test-cases.setup.js]: I use it to strip the file's extension properly. 
The equivalent transformation can be `resolvedEntry.replace('.code.js', '.tmp.js');`

Presumably, we can get rid off the `*.tmp.js` files with memoryFs, but It was a challenge for to debug it in the past :)

File: .eslintrc (hunk 5 above)
Comment (sullenor) [inline .eslintrc]: Done: https://github.com/webpack-contrib/eslint-config-webpack/pull/9

Comment (d3viant0ne) [inline .eslintrc]: @sullenor - Available in `eslint-config-webpack@latest`

File: test/test-cases.setup.js (hunk 6 above)
Comment (michael-ciniawsky) [inline test/test-cases.setup.js]: I'm also struggling with how to setup tests _at best_ in a generic way for 'architectural' consistency (where possibel) across the org, don't mind me too much in the scope of this PR, the current is totally 👍

File: .eslintrc (hunk 5 above)
Comment (sullenor) [inline .eslintrc]: I updated deps with `yarn.lock`, now it works as intended.
